torch/lib/THPP/tensors/generic/THSTensor.cpp

```diff
@@ -0,0 +1,454 @@
+#ifndef THS_GENERIC_FILE
+#define THS_GENERIC_FILE "tensors/generic/THSTensor.cpp"
+#else
+
+template<>
+THSTensor<real>::THSTensor():
+  tensor(THSTensor_(new)())
+  {};
+
+template<>
+THSTensor<real>::THSTensor(THSRealTensor *wrapped):
+  tensor(wrapped)
+  {};
+
+template<>
+THSTensor<real>::~THSTensor() {
+  if (tensor)
+    THSTensor_(free)(tensor);
+}
+
+template<>
+auto THSTensor<real>::clone() const -> THSTensor* {
+  return new THSTensor(THSTensor_(newClone)(tensor));
+}
+
+template<>
+auto THSTensor<real>::clone_shallow() -> THSTensor* {
+  THSTensor_(retain)(tensor);
+  return new THSTensor(tensor);
+}
+
+template<>
+int THSTensor<real>::nDim() const {
+  return tensor->nDimension;
+}
+
+template<>
+auto THSTensor<real>::sizes() const -> long_range {
+  return std::vector<long>(tensor->size, tensor->size + tensor->nDimension);
+}
+
+template<>
+const long* THSTensor<real>::rawSizes() const {
+  return tensor->size;
+}
+
+template<>
+auto THSTensor<real>::strides() const -> long_range {
+  throw std::runtime_error("THSTensor::strides() not supported");
+}
+
+template<>
+const long* THSTensor<real>::rawStrides() const {
+  throw std::runtime_error("THSTensor::rawStrides() not supported");
+}
+
+template<>
+std::size_t THSTensor<real>::storageOffset() const {
+  throw std::runtime_error("THSTensor::storageOffset() not supported");
+}
+
+template<>
+std::size_t THSTensor<real>::elementSize() const {
+  return sizeof(real);
+}
+
+template<>
+long long THSTensor<real>::numel() const {
+  throw std::runtime_error("THSTensor::>::() not supported");
```

Comment:
nit: There are a couple of bad replacements in the following functions:
* `data` (both of them)
* `cdata` (both)